winpr/libwinpr/sysinfo/sysinfo.c

```diff
@@ -553,35 +544,78 @@ DWORD GetTickCount(void)
 
 ULONGLONG winpr_GetTickCount64(void)
 {
-	ULONGLONG ticks = 0;
-#if defined(__linux__)
-	struct timespec ts;
+	const UINT64 ns = winpr_GetTickCount64NS();
+	return WINPR_TIME_NS_TO_MS(ns);
+}
 
-	if (!clock_gettime(CLOCK_MONOTONIC_RAW, &ts))
-		ticks = (ts.tv_sec * 1000) + (ts.tv_nsec / 1000000);
+#endif
+
+UINT64 winpr_GetTickCount64NS(void)
+{
+	UINT64 ticks = 0;
+#if _POSIX_C_SOURCE >= 200112L
+	struct timespec ts = { 0 };
+
+	if (clock_gettime(CLOCK_MONOTONIC_RAW, &ts) == 0)
+		ticks = (ts.tv_sec * 1000000000ull) + ts.tv_nsec;
 
 #elif defined(_WIN32)
-	FILETIME ft;
-	ULARGE_INTEGER ul;
-	GetSystemTimeAsFileTime(&ft);
-	ul.LowPart = ft.dwLowDateTime;
-	ul.HighPart = ft.dwHighDateTime;
-	ticks = ul.QuadPart;
+	union
+	{
+		FILETIME ft;
+		ULARGE_INTEGER ul;
+	} t;
+
+	GetSystemTimeAsFileTime(&t.ft);
+
+	static ULARGE_INTEGER first = { 0 };
```

Comment:
What do we do that static `first` stuff?
This changes the behavior of `GetTickCount64` as it will always return `0` on its first invocation (which might be unexpected since this should theoretically not be possible).

The Linux implementation now might also return negative (overflowed) values if `gettimeofday` fails.

Reply:
hmm, true, thought it was application start, but it is system start.
problem is that `GetSystemTimeAsFileTime` value is much too large for the `GetTickCount` and we need some reasonable cutoff.

Reply:
ok, fixed under windows by using `QueryPerformanceFrequency` with `QueryPerformanceCounter`
the fallback to `gettimeofday` now remembers the `uptime` on first measurement and returns the difference to that point in time.